### PR TITLE
More validation for components

### PR DESCRIPTION
**Pull request by gislenius**

### CODE REVIEW

File: queries/modules/io/28/modules/xbrl/components.jq [hunk 1]
```diff
@@ -1061,11 +1061,84 @@ declare function components:validation-errors(
...
+) as string*
+{
+  let $presentation-network as object? := networks:networks-for-components-and-short-names(
+      $component,
+      $networks:PRESENTATION_NETWORK)
+  let $roots := $presentation-network.Trees[]
+  let $hypercube := hypercubes:hypercubes-for-components($component)[not $$.Name eq "xbrl28:ImpliedTable"]
+  return (
+      "Error: Presentation network is missing."[empty($presentation-network)],
+      "Warning: Implied hypercube."[count($hypercube) gt 1],
```
Comment (dknochen) [inline queries/modules/io/28/modules/xbrl/components.jq]: should be eq 0

File: queries/modules/io/28/modules/xbrl/components.jq [hunk 2]
```diff
@@ -1061,11 +1061,84 @@ declare function components:validation-errors(
...
+        let $name := $hypercube-report-element.Name
+        where exists($name) and not $name = ("xbrl:Concept", "xbrl:Entity", "xbrl:Period", "xbrl:Unit", "xbrl28:Archive", "fsa:FiscalPeriod", "fsa:FiscalYear", "fsa:FiscalPeriodType", "sec:FiscalPeriod", "sec:FiscalYear", "sec:FiscalPeriodType", "fsa:Submitted")
+        let $presentation-report-element := descendant-objects($roots)[$$.Name = $name]
+        return
+        (
+          if(empty($presentation-report-element))
+          then "Error: " || $name || " not found in presentation."
+          else (),
+          let $presentation-children := $presentation-report-element.To[].Name
+          let $hypercube-children := ($hypercube-report-element.Members[].Name, $hypercube-report-element.Aspects."xbrl:Concept".Members[].Name, keys($hypercube-report-element.Aspects)[not $$ = ("xbrl:Concept", "xbrl:Entity", "xbrl:Period", "xbrl:Unit", "xbrl28:Archive", "fsa:FiscalPeriod", "fsa:FiscalYear", "fsa:FiscalPeriodType", "sec:FiscalPeriod", "sec:FiscalYear", "sec:FiscalPeriodType", "fsa:Submitted")])
```
Comment (dknochen) [inline queries/modules/io/28/modules/xbrl/components.jq]: newlines much appreciated

File: queries/modules/io/28/modules/xbrl/components.jq [hunk 3]
```diff
@@ -1064,11 +1064,104 @@ declare function components:validation-errors(
...
+) as string*
+{
+  let $presentation-network as object? := networks:networks-for-components-and-short-names(
+      $component,
+      $networks:PRESENTATION_NETWORK)
+  let $roots := $presentation-network.Trees[]
+  let $hypercube := hypercubes:hypercubes-for-components($component)[not $$.Name eq "xbrl28:ImpliedTable"]
+  return (
+      "Error: Presentation network is missing."[empty($presentation-network)],
+      "Error: More than one hypercube."[count($hypercube) gt 1],
```
Comment (dknochen) [inline queries/modules/io/28/modules/xbrl/components.jq]: not sure if we should mark this as an error in the generic case. Maybe a warning in the generic case and an error in the sec case?

File: queries/modules/io/28/modules/xbrl/components.jq [hunk 4]
```diff
@@ -1064,11 +1064,104 @@ declare function components:validation-errors(
...
+            where not $hypercube-children = $presentation-child
+            return "Error:  " || $presentation-child || " missing under " || $name || " in hypercube."
+          )
+        )
+      else (),
+      if(exists($roots))
+      then
+        for $hypercube-report-element in descendant-objects($hypercube)
+        let $name := $hypercube-report-element.Name
+        where exists($name) and not $name = ("xbrl:Concept", "xbrl:Entity", "xbrl:Period", "xbrl:Unit", "xbrl28:Archive", "fsa:FiscalPeriod", "fsa:FiscalYear", "fsa:FiscalPeriodType", "sec:FiscalPeriod", "sec:FiscalYear", "sec:FiscalPeriodType", "fsa:Submitted")
```
Comment (dknochen) [inline queries/modules/io/28/modules/xbrl/components.jq]: newlines much appreciated

File: queries/modules/io/28/modules/xbrl/components.jq [hunk 5]
```diff
@@ -1064,11 +1064,104 @@ declare function components:validation-errors(
...
+     then
+       let $presentation-children := $roots.To[]
+       return (
+         ("Error: The top-level abstract has more than two children (" || string-join($presentation-children.Name, ", ") || ")")
+           [count($presentation-children) gt 2],
+         "Error: The top-level abstract has no children."
+           [empty($presentation-children)],
+         if(count($presentation-children) eq 2)
+         then (
+           ("Error: Top-level abstract has inconsistent children ("|| string-join($presentation-children.Kind, ", ") || ")")[not $presentation-children.Kind = "Hypercube" or not $presentation-children.Kind = "LineItems"],
```
Comment (dknochen) [inline queries/modules/io/28/modules/xbrl/components.jq]: newlines

File: queries/modules/io/28/modules/xbrl/components.jq [hunk 6]
```diff
@@ -1064,11 +1064,104 @@ declare function components:validation-errors(
...
+                                                              "sec:FiscalPeriod",
+                                                              "sec:FiscalYear",
+                                                              "sec:FiscalPeriodType",
+                                                              "fsa:Submitted")]
+          )
+          return
+          (
+            for $hypercube-child in $hypercube-children
+            where not $presentation-children = $hypercube-child
+            return "Error:  " || $hypercube-child || " missing under " || $name || " in presentation."
```
Comment (dknochen) [inline queries/modules/io/28/modules/xbrl/components.jq]: "Error:  " -> "Error: "
(one space less)

**Comment by dknochen** [comment]
@gislenius I like that this is continuously improving.. We will have a complete validation suite at some point :)

**Comment by gislenius** [comment]
@dknochen Thanks and back to you!

**Pull request merged**